### ISSUE TITLE
Finetune errors

### ISSUE THREAD

**Issue by Unrealluver** [haotian-liu/LLaVA#83]
Hi, thanks for your latest commit! I noticed that I could not fine-tune with my pretrained ckpt `/home/zhulianghui/ProjectC_ChatGPT/llava/checkpoints/llava-13b-pretrain-8x40g-a100-fsdp-vicuna-0501-2`. When I execute the command like this:
```
bash /client-tools/repair_A100.sh
source /opt/conda/bin/activate /home/zhulianghui/.conda/envs/llava_beta
torchrun --nnodes=1 --nproc_per_node=4 --master_port=25001 \
    llava/train/train_mem.py \
    --model_name_or_path /home/zhulianghui/ProjectC_ChatGPT/llava/checkpoints/llava-13b-pretrain-8x40g-a100-fsdp-vicuna-0501-2 \
    --data_path /share/project/datasets/LLaVA-Instruct-150K/llava_instruct_150k.json \
    --image_folder /share/project/datasets/MSCOCO/coco2017/train2017 \
    --vision_tower /share/project/wyz/MM/models/openai/clip-vit-large-patch14 \
    --pretrain_mm_mlp_adapter /home/zhulianghui/ProjectC_ChatGPT/llava/checkpoints/llava-13b-pretrain-8x40g-a100-fsdp-vicuna-0501-2/llava-13b-pretrain.bin \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end True \
    --output_dir ./checkpoints/llava-13b-finetune \
    --num_train_epochs 3 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 8 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 5000 \
    --save_total_limit 3 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --bf16 True \
    --fsdp "full_shard auto_wrap offload" \
    --fsdp_transformer_layer_cls_to_wrap 'LlamaDecoderLayer' \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --run_name "vicuna-13b-clip-l-finetune-llava" \
    --lazy_preprocess True \
    --report_to mlflow
```
I got the error below:
![image](https://user-images.githubusercontent.com/27504120/235639600-69e0d5f8-66ef-49ad-a07f-5bc6c0983a50.png)

I checked the `llava/model/llava.py` and set several logs like below:
```
        if mm_use_im_start_end:
            num_new_tokens = tokenizer.add_tokens([DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN], special_tokens=True)
            print(f"Added {num_new_tokens} tokens to the tokenizer.")
            self.resize_token_embeddings(len(tokenizer))
            vision_config.im_start_token, vision_config.im_end_token = tokenizer.convert_tokens_to_ids([DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN])
            print(f"Set im_start_token to {vision_config.im_start_token} and im_end_token to {vision_config.im_end_token}.")

```

the `num_new_tokens` is 0, which means that the tokenizer already contains the `DEFAULT_IM_START_TOKEN` and `DEFAULT_IM_END_TOKEN`. So the code below will face some errors: 1. `input_embeddings` referenced before the assignment. 2. num_new_tokens != 2 

```
            if pretrain_mm_mlp_adapter:
                mm_projector_weights = torch.load(pretrain_mm_mlp_adapter, map_location='cpu')
                embed_tokens_weight = mm_projector_weights['model.embed_tokens.weight']
                assert input_embeddings.shape == embed_tokens_weight.shape
                assert num_new_tokens == 2
                input_embeddings[-num_new_tokens:] = embed_tokens_weight[-num_new_tokens:]

```
So, I change the codes to these:
```
            if pretrain_mm_mlp_adapter:
                if num_new_tokens > 0:
                    mm_projector_weights = torch.load(pretrain_mm_mlp_adapter, map_location='cpu')
                    embed_tokens_weight = mm_projector_weights['model.embed_tokens.weight']
                    assert input_embeddings.shape == embed_tokens_weight.shape
                    assert num_new_tokens == 2
                    input_embeddings[-num_new_tokens:] = embed_tokens_weight[-num_new_tokens:]
                else:
                    input_embeddings = self.get_input_embeddings().weight.data
                    mm_projector_weights = torch.load(pretrain_mm_mlp_adapter, map_location='cpu')
                    embed_tokens_weight = mm_projector_weights['model.embed_tokens.weight']
                    assert (input_embeddings == embed_tokens_weight).all()
```
It could continue to fine-tune. I wonder if the changes lead to a wrong fine-tuning process. I would be appreciate if you could give me some guidance!

**Comment by haotian-liu** [comment]
Hi @Unrealluver, maybe there is some ambiguity in the way I name my parameter arguments.

`--pretrain_mm_mlp_adapter /home/zhulianghui/ProjectC_ChatGPT/llava/checkpoints/llava-13b-pretrain-8x40g-a100-fsdp-vicuna-0501-2/llava-13b-pretrain.bin` loads the pretrained projector from this file `args.pretrain_mm_mlp_adapter`, that is extracted using `scripts/extract_projector.py`.
The based checkpoint you loaded is `--model_name_or_path /home/zhulianghui/ProjectC_ChatGPT/llava/checkpoints/llava-13b-pretrain-8x40g-a100-fsdp-vicuna-0501-2`, also seems to be some LLaVA-pretrained/finetuned checkpoints.

This suggests that either the projector weights are contained in both `args.model_name_or_path` and `args.pretrain_mm_mlp_adapter`.  It will either be redundant, or there would be a conflict.

If you would like to finetune the projector only, you can add `--tune_mm_mlp_adapter`.
If the projector weights in `args.model_name_or_path` and `args.pretrain_mm_mlp_adapter` is the same, then you can either (1) do not use `--pretrain_mm_mlp_adapter` at all, or (2) set `args.model_name_or_path` to `Vicuna` weights, if there are no other weights tuned.

`args.pretrain_mm_mlp_adapter` is used so that (1) we can share to public the pretrained projectors (first stage in our case) without downloading the untuned Vicuna weights. (2) you can share pretrained projector easily between servers for the second stage instruction tuning -- it's much smaller than the full model weights.  If you are using the same server to train/finetune your checkpoint, you may simply choose not using this :)

Please let me know if there are still things that's unclear.

**Comment by Unrealluver** [comment]
Thanks for your detailed answer! I will try it again.

**Comment by Unrealluver** [comment]
Thanks for your help in solving this problem. I still have some questions about the checkpoint saved by `LLaVA pretrain (step1)`

Here are the contents of `vicuna-13b-v0`:
![image](https://user-images.githubusercontent.com/27504120/236236578-33d27608-8ddd-49d3-85db-38f9ae7a74df.png)

And here are the contents of checkpoints saved by `LLaVA pretrain (step1)`
![image](https://user-images.githubusercontent.com/27504120/236236953-8fb55dd4-67e9-4264-9bd4-600f62127ad7.png)

My questions are as follows:
(1) If the contents of checkpoints saved by `LLaVA pretrain (step1)` contains parameters of `vicuna-13b-v0`, `projection layer`, and `CLIP-L`?
(2) If the answer to (1) is yes. If the `vicuna-13b-v0` and `LLaVA pretrain (step1)` contains the same weights for the language part?
(3) If the `LLaVA pretrain (step1)` contains the same weights as `CLIP-L` for the vision part expect for the `projection layer`?
(4) When the `--tune_mm_mlp_adapter` is `True`, why do the codes in `llava/model/llava.py` work as bellow:
```
                        if orig_embeds_params is not None:
                            cur_new_input_embeds = torch.cat((cur_input_embeds[:image_start_token_pos].detach(), cur_input_embeds[image_start_token_pos:image_start_token_pos+1], cur_image_features, cur_input_embeds[image_start_token_pos + num_patches + 1:image_start_token_pos + num_patches + 2], cur_input_embeds[image_start_token_pos + num_patches + 2:].detach()), dim=0)
                        else:
                            cur_new_input_embeds = torch.cat((cur_input_embeds[:image_start_token_pos+1], cur_image_features, cur_input_embeds[image_start_token_pos + num_patches + 1:]), dim=0)
```

The `:image_start_token_pos+1` is strange to me. Why we should add `.detach()` here? I think we have set the whole model except the `projection layer` to `required_grad = False` already. The long codes here with long indices puzzle me a lot. I will appreciate it if you could give me some guidance here.

Thanks for your time!

**Comment by yunh-w** [comment]
I meet the same error.

**Comment by haotian-liu** [comment]
Hi @Unrealluver,

Regarding your questions
(1) it contains vicuna-13b-v0 + projection layer (and potentially additional token embeddings for <im_start> and <im_end> etc), it does not contain CLIP-L.
(2) yes, so that we have this [extract_projector.py](https://github.com/haotian-liu/LLaVA/blob/main/scripts/extract_mm_projector.py) that allow you to only share the projectors and do not need to share other weights.
(3) yes.
(4) This is for tuning the `<im_start>` and `<im_end>` only, and for other token embeddings, it will be detached and not updated.

Please let me know if there are anything unclear, thanks.

**Comment by Coolshanlan** [comment]
Thank you so much for your quick reply and modify, it worked perfectly!!
